diff --git a/src/ap.c b/src/ap.c
index 8ac1ede..f17a485 100644
--- a/src/ap.c
+++ b/src/ap.c
@@ -3315,13 +3315,15 @@ static struct l_dbus_message *ap_dbus_stop(struct l_dbus *dbus,
 	return NULL;
 }
 
+void cleanup_config(void *ptr) { l_settings_free(*(void **) ptr); }
 static struct l_dbus_message *ap_dbus_start_profile(struct l_dbus *dbus,
 						struct l_dbus_message *message,
 						void *user_data)
 {
 	struct ap_if_data *ap_if = user_data;
 	const char *ssid;
-	_auto_(l_settings_free) struct l_settings *config = NULL;
+	__attribute__((cleanup(cleanup_config)))
+		struct l_settings *config = NULL;
 	char *config_path;
 	int err;
 
diff --git a/src/network.c b/src/network.c
index 6d569b3..caae4f9 100644
--- a/src/network.c
+++ b/src/network.c
@@ -375,6 +375,7 @@ static bool network_set_8021x_secrets(struct network *network)
 	return true;
 }
 
+void cleanup_network(void *ptr) { l_free(*(void **) ptr); }
 static int network_load_psk(struct network *network, bool need_passphrase)
 {
 	const char *ssid = network_get_ssid(network);
@@ -382,10 +383,10 @@ static int network_load_psk(struct network *network, bool need_passphrase)
 	size_t psk_len;
 	uint8_t *psk = l_settings_get_bytes(network->settings, "Security",
 						"PreSharedKey", &psk_len);
-	_auto_(l_free) char *passphrase =
+	__attribute__((cleanup(cleanup_network))) char *passphrase =
 			l_settings_get_string(network->settings,
 						"Security", "Passphrase");
-	_auto_(l_free) char *path =
+	__attribute__((cleanup(cleanup_network))) char *path =
 		storage_get_network_file_path(security, ssid);
 	int r;
 
