diff --git a/src/eap.c b/src/eap.c
index bef46e0..5c8050f 100644
--- a/src/eap.c
+++ b/src/eap.c
@@ -389,6 +389,16 @@ static const char *eap_type_to_str(enum eap_type type, uint32_t vendor_id,
 #define IS_EXPANDED_RESPONSE(id, t) \
 	(type == EAP_TYPE_EXPANDED && vendor_id == (id) && vendor_type == (t))
 
+#define response_is(W) _response_is(type, (W), vendor_id, vendor_type)
+bool _response_is(enum eap_type type, enum eap_type wanted,
+		uint32_t vendor_id, uint32_t vendor_type)
+{
+	if (type == wanted)
+		return true;
+
+	return IS_EXPANDED_RESPONSE(0, wanted);
+}
+
 static void eap_handle_response(struct eap_state *eap, const uint8_t *pkt,
 				size_t len)
 {
@@ -401,14 +411,6 @@ static void eap_handle_response(struct eap_state *eap, const uint8_t *pkt,
 				eap->method->vendor_id[2];
 	uint32_t our_vendor_type = eap->method->vendor_type;
 
-	bool response_is(enum eap_type wanted)
-	{
-		if (type == wanted)
-			return true;
-
-		return IS_EXPANDED_RESPONSE(0, wanted);
-	}
-
 	if (len < 1)
 		/* Invalid packets to be ignored */
 		return;

